### ISSUE TITLE
Proposal for subtitle support and other changes in convert-video

### ISSUE THREAD

**Issue by donmelton** [donmelton/video_transcoding#114]
**Update:** This proposal has been updated to reflect a change in the name and behavior of its new command line option.

# Proposal for subtitle support and other changes in `convert-video`

While the rest of the tools in my `video_transcoding` Gem received significant new features, bug fixes and other updates during 2016, the `convert-video` tool was left virtually unchanged.

It's 2017 now and time to give the `convert-video` tool some love. By re-writing it. :)

The `transcode-video` tool has always supported adding subtitles, but `convert-video` has never included them in its output during conversion. This was partly due to the lack of support from other tools for proper conversion between text-based subtitle formats.

But it was also a design decision by me to try and convince users to stop adding subtitle tracks in their original format to their output files.

To remind everyone why including image-based subtitles in your transcodings is an idea apparently inspired by Satan, here's what I say in the "[Understanding subtitles](https://github.com/donmelton/video_transcoding#understanding-subtitles)" section of the "README" document about it:

> Blu-ray Disc and DVD subtitles are bitmap formats. They're not text. They're large, unwieldy and may not appear correctly if you crop your video. Blu-ray Disc-format subtitles aren't even allowed in MP4 output. And DVD-format subtitles, while allowed, often won't display at all in many MP4 players.

Of course, this didn't stop some of you from doing it anyway. :)

To be clear, I have no problem with text-based subtitles. But this change I'm proposing to `convert-video` will actually include support for the evil kind of subtitles too. Why? Because it's easy to implement once you write all the infrastructure to support text-based conversion.

The original implementation of `convert-video` used two main tools for the conversion process. The `ffmpeg` tool was used to convert from from Matroska (MKV) to MP4 format because audio would sometimes need to be transcoded during the process. However, `mkvmerge` was used to convert from MP4 to MKV format because the process itself was just a matter of re-muxing the streams.

The new implementation will use `ffmpeg` for both directions of the process because `mkvmerge` can't handle conversion from SubRip subtitle format, used in MKV files, back to 3GPP Timed Text subtitle format, used in MP4 format.

It would be charitable to describe the handling of metadata by `ffmpeg` as "maddening." Which is why the `mp4track` tool has always been used to enable and disable tracks as necessary for MP4 output. The new implementation will, much like `transcode-video` already does, leverage the `mkvpropedit` tool for similar duties with MKV files.

Which means that while `convert-video` will no longer depend on `mkvmerge`, a dependency on `mkvpropedit` will be added. Still, it will reduce the overall number of dependencies for the whole `video_transcoding` Gem. And let me remove an entire source file from the codebase.

There's a good chance that `ffmpeg` will be able to handle translation from other text-based subtitles found within rogue MKV files, e.g. Advanced SubStation Alpha, WebVTT or even Raw format, to the 3GPP Timed Text format in MP4s. It's also possible that Raw text format in rogue MP4 files might be handled correctly.

And `ffmpeg` can copy DVD-style VobSub format image-based subtitles unchanged between MKV and MP4 files. But Blu-ray Disc-style PGS format image-based subtitles won't be allowed.

There may be some other side effects from using `ffmpeg` for handling conversion to MKV format. For example, `ffmpeg` appears to copy over some non-essential MP4 attributes to the target MKV file as slightly non-standard metadata. And it leaves out some other metadata like the non-essential language attribute on chapters that `mkvmerge` used to add. Go figure. But none of these differences affect playback in any way.

Included with all this subtitle support will be four other changes.

The first is a new option called `--no-double` that will allow users, like me, to prevent `convert-video` from assuming input files might contain two main audio tracks whose order needs to be swapped or that a "missing" stereo AAC audio track needs to be added to MP4 output. Here's how it will be described in the built-in help:

```
    --no-double     don't swap order of first two audio tracks
                      or add stereo copy of main surround track
```

The second change is to the algoritm deciding when that "missing" stereo AAC audio track is added. Previously that only happened when the first track of the input MKV file was in surround format _and_ there were no other audio tracks. Now it won't matter how many audio tracks are in the input.

The third change is simply to increase the bitrate of any AC-3 tracks created in MP4 files from 384 Kbps to 640 Kbps, the same default now used in `transcode-video`.

The remaining change has to do with requiring a new, modern version of `ffmpeg` to do all this subtitle conversion foo. From now on, only the built-in, native FFmpeg AAC encoder and the optional Fraunhofer FDK AAC encoder will be supported.

The bad news is that because of the stupid way that versioning works in the FFmpeg world, I can't reliably check for which one you have installed. Seriously, the code will just fail in mysterious and unpredictable ways using an old `ffmpeg` binary. #FML

The good news is that it's easy to get a new, modern version of `ffmpeg`. If you're on macOS and use Homebrew to keep your dependencies up to date then you're good. If you're on Windows or Linux then downloads are available even if your package management system doesn't have the latest.

The code for this entire "proposal" is actually complete. Thank god for shitty weather this weekend because I was able to finish it all last night.

And it works! Of course I'll be testing it much more in the next few days, but the hard part will be exploring the "edges," as they say. And to do that, I'll need your help. Because even I don't have enough whacky samples to run through conversion. But you probably have some interesting test cases.

So, it's likely I'll be checking all of this in and releasing it sometime this week. And then we can just open new issues for any "oops" cases that I might have missed.

Let me know what you think. Soon.

Thanks.

**Comment by donmelton** [comment]
OK, I released this as version 0.15.0. Just `gem update video_transcoding` to get it. Thanks.

**Comment by platonium** [comment]
Much obliged!

On Jan 15, 2017, at 11:43 AM, Don Melton <notifications@github.com<mailto:notifications@github.com>> wrote:


OK, I released this as version 0.15.0. Just gem update video_transcoding to get it. Thanks.
